### PR TITLE
2 fa

### PR DESCRIPTION
**Pull request by kolesnikdi**
some 2fa shit

### CODE REVIEW

File: Web_Menu_DA/settings.py [hunk 1]
```diff
@@ -199,6 +200,7 @@
     'default': {
         'BACKEND': 'django_redis.cache.RedisCache',
         'LOCATION': 'redis://redis_app:6379/1',
+        'LOCATION': 'redis://localhost:6379/1',     # need for work on local machine (menu view)
```
Comment (ArtemKhriapa) [inline Web_Menu_DA/settings.py:203]: fixme

File: Web_Menu_DA/settings.py [hunk 2]
```diff
@@ -209,7 +211,15 @@
...
-        'functional': 60 * 2 * 2, }
+        'functional': 60 * 2 * 2, },
+    'CompanyViewSet': {
+        '2fa': 60 * 5,
+    }
+}
+
+CACHE_TIMEOUT_2FA = {
+    'email':  60 * 5,
+    'google':  60 * 5,
```
Comment (ArtemKhriapa) [inline Web_Menu_DA/settings.py:222]: fixme

File: company/views.py [hunk 3]
```diff
@@ -2,14 +2,21 @@
...
 from Web_Menu_DA.permissions import IsOwnerOr404
 
+from two_factor_authentication.business_logic import enabled_2fa
+
 
 class CompanyViewSet(viewsets.ModelViewSet):
     permission_classes = [IsAuthenticated, IsOwnerOr404]
     serializer_class = CompanySerializer
 
+    @enabled_2fa()  # reqieride = True
```
Comment (ArtemKhriapa) [inline company/views.py:16]: required

File: menu/views.py [hunk 4]
```diff
@@ -11,33 +11,34 @@
 from menu.filters import CustomSearchFilter, CustomRangeFilter, perform_product_filtering
 from menu.serializers import LocationMenuSerializer, ProductMenuSerializer
 
-
 TIMEOUT = CACHE_TIMEOUT.get('LocationMenuView')
 
+
+@method_decorator(cache_page(TIMEOUT['decorator']), name='retrieve')
```
Comment (ArtemKhriapa) [inline menu/views.py:17]: # it's able to use decorator for class with method name as parameter

File: two_factor_authentication/business_logic.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+from rest_framework.response import Response
+
+from Web_Menu_DA.settings import CACHE_TIMEOUT_2FA
+
+
+# class AUTH2TYPES(enum.IntEnum): # MOVE IT INTO USER CONSTANTS # todo when implement Google Auth/ move it to constants
+#     NONE = 0
+#     EMAIL = 1
+#     SMS = 2
+#     GAUTH = 3
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:27]: move into constants!

File: two_factor_authentication/business_logic.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        return cls._check_2fa(user, code)
+
+    @classmethod
+    def _check_2fa(cls, user, code):
+        raise NotImplementedError
+
+
+class Cache2FA(Base2FA):
+    """This class is intended for followers that use the cache to write 2fa code """
+
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:98]: cache_delay = 300

File: two_factor_authentication/business_logic.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    @classmethod
+    def cache_set(cls, user):
+        code_2fa = ''.join(random.choices('123456789', k=6))
+        # cache.set(user.id, {'code': code_2fa}, CACHE_TIMEOUT_2FA.get(user.type_2fa))
+        cache.set(user.id, {'code': code_2fa}, CACHE_TIMEOUT_2FA.get('email'))
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:103]: cache.set(user.id, {'code': code_2fa}, cls. cache_delay)

File: two_factor_authentication/business_logic.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        return code_2fa
+
+
+class Auth2FAEmail(Cache2FA):
+    """Class which uses Email for checking or performing Two-factor verification"""
+    auth2fa_type = 'Email'
+
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:110]: cache_delay = CACHE_TIMEOUT_2FA.get('email')

File: Web_Menu_DA/settings.py (hunk 1 above)
Comment (kolesnikdi) [reply Web_Menu_DA/settings.py:203]: done

File: Web_Menu_DA/settings.py (hunk 2 above)
Comment (kolesnikdi) [reply Web_Menu_DA/settings.py:222]: done

File: company/views.py (hunk 3 above)
Comment (kolesnikdi) [reply company/views.py:16]: done

File: menu/views.py (hunk 4 above)
Comment (kolesnikdi) [reply menu/views.py:17]: done

File: two_factor_authentication/business_logic.py (hunk 5 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:27]: done

File: two_factor_authentication/business_logic.py (hunk 8 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:110]: done

File: two_factor_authentication/business_logic.py (hunk 7 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:103]: done

File: two_factor_authentication/business_logic.py (hunk 6 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:98]: done

**Review by ArtemKhriapa — APPROVED** [review]

**Pull request merged**